cmd/zpool/zpool_iter.c

```diff
@@ -316,3 +318,162 @@ for_each_vdev(zpool_handle_t *zhp, pool_vdev_iter_f func, void *data)
 	}
 	return (for_each_vdev_cb(zhp, nvroot, func, data));
 }
+
+/* Thread function run for each vdev */
+static void
+vdev_run_cmd_thread(void *cb_cmd_data)
+{
+	vdev_cmd_data_t *data = cb_cmd_data;
+
+	char *pos = NULL;
+	FILE *fp;
+	size_t len = 0;
+	char cmd[MAXPATHLEN];
+
+	/* Set our VDEV_PATH and VDEV_UPATH env vars and run command */
+	snprintf(cmd, sizeof (cmd), "VDEV_PATH=%s && VDEV_UPATH=%s && %s",
+	    data->path, data->upath ? data->upath : "\"\"", data->cmd);
+
+	fp = popen(cmd, "r");
+	if (!fp)
+		return;
+
+	data->line = NULL;
+
+	/* Save the first line of output from the command */
+	if (getline(&data->line, &len, fp) != -1) {
+		/* Success.  Remove newline from the end, if necessary. */
+		if ((pos = strchr(data->line, '\n')) != NULL)
+		    *pos = '\0';
+	} else {
+		data->line = NULL;
+	}
+	pclose(fp);
+}
+
+/* For each vdev in the pool run a command */
+static int
+for_each_vdev_run_cb(zpool_handle_t *zhp, nvlist_t *nv, void *cb_vcdl)
+{
+	vdev_cmd_data_list_t *vcdl = cb_vcdl;
+	vdev_cmd_data_t *data;
+	char *path = NULL;
+	int rc = 1;
+	int i;
+
+	if (nvlist_lookup_string(nv, ZPOOL_CONFIG_PATH, &path) != 0 || !path)
+		goto end;
+
+	/* Spares show more than once if they're in use, so skip if exists */
+	for (i = 0; i < vcdl->count; i++) {
+		if ((strcmp(vcdl->data[i].path, path) == 0) &&
+		    (strcmp(vcdl->data[i].pool, zpool_get_name(zhp)) == 0)) {
+			/* vdev already exists, skip it */
+			rc = 0;
+			goto end;
+		}
+	}
+
+	/*
+	 * Resize our array and add in the new element.
+	 */
+	if (!(vcdl->data = realloc(vcdl->data,
+	    sizeof (*vcdl->data) * (vcdl->count + 1))))
+		goto end;	/* couldn't realloc */
```

Comment:
I'd rather keep the realloc code simple than try to optimize it for speed.  The realloc time is going to be insignificant in comparison to the runtime of the command we spawn.  If it really is a performance bottleneck, we can always change it.